### ISSUE TITLE
[Proposal] innerHTML Security

### ISSUE THREAD

**Issue by EisenbergEffect** [microsoft/fast-dna#2850]
## Description

In FastElement, it's possible to bind to `innerHTML` using the following template syntax:

```HTML
<div :innerHTML=${x => x.someDangerousHTMLContent}></div>
```

This opens up the door to various security issues. We can rely on developers to do the right thing, by sanitizing input, as strongly suggested in our documentation:

```HTML
<div :innerHTML=${x => sanitize(x.someDangerousHTMLContent)}></div>
```

However, this still relies on developers to know that they need to do this, and it's quite possible that they will make a mistake or forget. To address this issue, I suggest that we detect bindings against `innerHTML` during template compilation, and wrap the expression that retrieves the content with the new TrustedTypes policy feature. By default, the FastElement library would create a policy that throws an error, requesting that customers configure a real policy with their sanitization rules. In this way, we can guarantee that XSS attacks are not possible through `innerHTML` bindings and we also can hook into the browser's CSP features (which allow policy names to be locked down).

## Examples

Looking back at the first, problematic example:

```HTML
<div :innerHTML=${x => x.someDangerousHTMLContent}></div>
```

During template compilation, the expression would be transformed from this:

```TypeScript
x => x.someDangerousHTMLContent
```

To this:

```TypeScript
x => fastHTMLPolicy.createHTML(x.someDangerousHTMLContent)
```

By making this transform, we ensure that all html bindings go through our policy. Again, by default, the policy would throw, requesting the developer to configure a real policy.

To configure a policy, we'll Likely have an API like this:

```TypeScript
const htmlPolicy = trustedTypes.createPolicy('my-policy', {
  createHTML: (html) => {
    // TODO: call some sanitization library on the html
    return html;
  }
});

BindingDirective.configureHTMLPolicy(htmlPolicy);
```

Under the hood, the BindingDirective would have a default implementation like this:

```TypeScript
let fastHTMLPolicy = trustedTypes.createPolicy('fast-html-policy', {
  createHTML: (html) => {
    throw new Error('Please configure an HTML policy..blah..blah..');
  }
});
```

## Dependencies

Implementing the feature in this way relies on the TrustedTypes feature being available in browsers. It is not yet ready, but is behind a flag in Chromium. However, a tiny polyfill, which enables the API to work can be accomplished with this code:

```TypeScript
if(typeof trustedTypes == 'undefined')trustedTypes={createPolicy:(n, rules) => rules};
```

More functional, full polyfills, are also available.

## Resources

* [Trusted Types W3C Spec](https://w3c.github.io/webappsec-trusted-types/dist/spec/)
* [Trusted Types Polyfills](https://github.com/w3c/webappsec-trusted-types)
* [Google Developer Blog](https://developers.google.com/web/updates/2019/02/trusted-types)

**Comment by nicholasrice** [comment]
I'm not super familiar with this API - is there a case where a developer might different policies for different components? Is a single `configure` API sufficient or would we need support for instance policies?

**Comment by EisenbergEffect** [comment]
The trusted types API allows the creation of various policies which can be controlled through an HTTP header. For example, this header:

```
Content-Security-Policy: trusted-types my-policy
```

With this in place, code cannot set innerHTML without using `my-policy` to do so. Since setting the policy is up to the person or team that creates the app, I think that we should not have a per-component or per-binding policy. This opens up lots of security holes. Furthermore, I don't think our default policy should be allowed to be set more than once. Malicious code could attempt to replace the policy if that were the case.

**Comment by EisenbergEffect** [comment]
Implemented in #2855
